avalanche/training/regularization.py

```diff
@@ -166,4 +166,157 @@ def __call__(self, out_in, target_in, out_buffer, target_buffer):
+        is_invalid[invalid_idx] = 1
+        if invalid_idx.sum() > 0:
+            # avoid operand fail
+            valid_pos[:, invalid_idx] = 1
+            valid_neg[:, invalid_idx] = 1
+
+        pos_idx = torch.multinomial(valid_pos.float().T, 1).squeeze(1)
+        neg_idx = torch.multinomial(valid_neg.float().T, 1).squeeze(1)
+
+        pos_x = x_all[pos_idx]
+        pos_y = y_all[pos_idx]
+        neg_x = x_all[neg_idx]
+        neg_y = y_all[neg_idx]
+
+        return pos_x, pos_y, neg_x, neg_y, is_invalid
+
+    def __sup_con_loss(
+        self,
+        anchor_features: torch.Tensor,
+        features: torch.Tensor,
+        anchor_targets: torch.Tensor,
+        targets: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Method able to compute the supervised contrastive loss of new minibatch.
+        @param anchor_features: Anchor features related to new minibatch duplicated mapped in latent space.
+        @param features: Features related to half positive and half negative examples mapped in latent space.
+        @param anchor_targets: Labels related to anchor features.
+        @param targets: Labels related to features.
+        @return: Supervised contrastive loss.
+        """
+        pos_mask = (
+            (anchor_targets.reshape(-1, 1) == targets.reshape(1, -1))
+            .float()
+            .to(self.device)
+        )
+        similarity = anchor_features @ features.T / self.temp
+        similarity -= similarity.max(dim=1)[0].detach()
+        log_prob = similarity - torch.log(torch.exp(similarity).sum(1))
+        mean_log_prob_pos = (pos_mask * log_prob).sum(1) / pos_mask.sum(1)
+        loss = -(self.temp / self.base_temp) * mean_log_prob_pos.mean()
+        return loss
+
+    def __scale_by_norm(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Function able to scale by its norm a certain tensor.
+        @param x: Tensor to normalize.
+        @return: Normalized tensor.
+        """
+        x_norm = torch.norm(x, p=2, dim=1).unsqueeze(1).expand_as(x)
+        return x / (x_norm + 1e-05)
+
+    def __call__(
+        self,
+        input_in: torch.Tensor,
+        target_in: torch.Tensor,
+        output_buffer: torch.Tensor,
+        target_buffer: torch.Tensor,
+        buffer_replay_data: tuple,
+    ) -> torch.Tensor:
+        """
+        Method able to compute the ER_AML loss.
+        @param input_in: New inputs examples.
+        @param target_in: Labels of new examples.
+        @param output_buffer: Predictions of samples from buffer.
+        @param target_buffer: Labels of samples from buffer.
+        @param buffer_replay_data: Buffer replay data to compute positive and negative samples.
+        @return: ER_AML computed loss.
+        """
+        x_buffer, y_buffer, _ = buffer_replay_data
+        pos_x, pos_y, neg_x, neg_y, is_invalid = self.__sample_pos_neg(
+            input_in, target_in, x_buffer, y_buffer
+        )
```

Comment:
I tried to use a batch instead of the total memory for sampling positives and negatives and results are:
- cifar10: 0.48
- cifar100: 0.21

instead with the whole memory results are:
- cifar10: 0.49 accuracy
- cifar100: 0.25 accuracy